File: Zen-pos2/src/api/locations.ts
import { apiRequest } from './client';

export interface Location {
  id: string;
  name: string;
  subtitle: string;
  address: string;
  phone: string;
  email: string;
  tablesCount: number;
  barCount: number;
  isActive: boolean;
  openingTime?: string;
  closingTime?: string;
}

interface ApiLocation {
  id: string;
  name: string;
  subtitle: string;
  address: string;
  phone: string;
  email: string;
  tables_count: number;
  bar_count: number;
  is_active: boolean;
  opening_time?: string;
  closing_time?: string;
}

function mapLocation(raw: ApiLocation): Location {
  return {
    id: raw.id,
    name: raw.name,
    subtitle: raw.subtitle ?? '',
    address: raw.address,
    phone: raw.phone,
    email: raw.email,
    tablesCount: raw.tables_count,
    barCount: raw.bar_count,
    isActive: raw.is_active,
    openingTime: raw.opening_time,
    closingTime: raw.closing_time,
  };
}

export async function listLocations(): Promise<Location[]> {
  const raw = await apiRequest<ApiLocation[]>('/locations/');
  return raw.map(mapLocation);
}

export async function createLocation(payload: Omit<Location, 'id' | 'isActive'>): Promise<Location> {
  const raw = await apiRequest<ApiLocation>('/locations/', {
    method: 'POST',
    body: JSON.stringify({
      name: payload.name,
      subtitle: payload.subtitle,
      address: payload.address,
      phone: payload.phone,
      email: payload.email,
      tables_count: payload.tablesCount,
      bar_count: payload.barCount,
      opening_time: payload.openingTime,
      closing_time: payload.closingTime,
    }),
  });
  return mapLocation(raw);
}

export async function updateLocation(id: string, payload: Partial<Omit<Location, 'id' | 'isActive'>>): Promise<Location> {
  const raw = await apiRequest<ApiLocation>(`/locations/${id}`, {
    method: 'PUT',
    body: JSON.stringify({
      ...(payload.name !== undefined && { name: payload.name }),
      ...(payload.subtitle !== undefined && { subtitle: payload.subtitle }),
      ...(payload.address !== undefined && { address: payload.address }),
      ...(payload.phone !== undefined && { phone: payload.phone }),
      ...(payload.email !== undefined && { email: payload.email }),
      ...(payload.tablesCount !== undefined && { tables_count: payload.tablesCount }),
      ...(payload.barCount !== undefined && { bar_count: payload.barCount }),
      ...(payload.openingTime !== undefined && { opening_time: payload.openingTime }),
      ...(payload.closingTime !== undefined && { closing_time: payload.closingTime }),
    }),
  });
  return mapLocation(raw);
}

export async function deleteLocation(id: string): Promise<void> {
  await apiRequest<void>(`/locations/${id}`, { method: 'DELETE' });
}

File: zen-pos-api/app/routers/locations.py
from fastapi import APIRouter, Depends

from app.dependencies import require_permission
from app.models.location import LocationDocument
from app.schemas.location import LocationCreate, LocationUpdate, LocationOut
from app.core.exceptions import NotFoundError

router = APIRouter()


@router.get("/", response_model=list[LocationOut])
async def list_locations():
    """List all active locations. No auth required — needed for onboarding dropdowns."""
    locations = await LocationDocument.find(LocationDocument.is_active == True).to_list()  # noqa: E712
    return [_to_out(loc) for loc in locations]


@router.get("/{location_id}", response_model=LocationOut)
async def get_location(location_id: str):
    loc = await LocationDocument.get(location_id)
    if not loc:
        raise NotFoundError("Location not found")
    return _to_out(loc)


@router.post("/", response_model=LocationOut, status_code=201,
             dependencies=[Depends(require_permission("view_settings"))])
async def create_location(body: LocationCreate):
    loc = LocationDocument(**body.model_dump())
    await loc.insert()
    return _to_out(loc)


@router.put("/{location_id}", response_model=LocationOut,
            dependencies=[Depends(require_permission("view_settings"))])
async def update_location(location_id: str, body: LocationUpdate):
    loc = await LocationDocument.get(location_id)
    if not loc:
        raise NotFoundError("Location not found")
    for key, value in body.model_dump(exclude_unset=True).items():
        setattr(loc, key, value)
    await loc.save()
    return _to_out(loc)


@router.delete("/{location_id}", status_code=204,
               dependencies=[Depends(require_permission("view_settings"))])
async def delete_location(location_id: str):
    loc = await LocationDocument.get(location_id)
    if not loc:
        raise NotFoundError("Location not found")
    loc.is_active = False
    await loc.save()


def _to_out(loc: LocationDocument) -> LocationOut:
    return LocationOut(
        id=str(loc.id),
        name=loc.name,
        subtitle=loc.subtitle,
        address=loc.address,
        phone=loc.phone,
        email=loc.email,
        tables_count=loc.tables_count,
        bar_count=loc.bar_count,
        is_active=loc.is_active,
    )
